### PR TITLE
Delegate V3 Admin to Admin ACL contract

### PR DESCRIPTION
**Pull request by ryley-o**
Delegate all V3 admin authentication to a generalized Admin ACL contract, and assume Admin ACL is core contract owner.

This implementation is based on discussions between @ryley-o, @jakerockland, and @laureli(https://github.com/laureli).

In general, whenever admin may have authority to call some external function on the core contract, the core contract asks its owning ACL "is `msg.sender` allowed to call the external function being called on `address(this)`?". The ACL responds with true or false, and the core contract either rejects or accepts the call based on the response.

It is expected that eventually there will be some AdminACL contract designed that utilizes groups and sets of addresses/multisigs to determine which groups may call which functions on the core contract.

The details of this implementation are somewhat elegant in that function interfaces are relatively easy to generate in solidity, resulting in fairly readable code. This is also true in ethers, making tests fairly straightforward.

## Design Alternatives
Delegating all calls to an AdminACL (i.e. requiring all admin calls have `msg.sender == owner()`) was considered, specifically because it would be backwards compatible with V0 and V1 Art Blocks core contracts. However, if generalized, that pattern would require admin EOAs and/or multisigs to generate raw calldata in some way, and interact with the AdminACL contract directly.

Perhaps the most competitive and still relatively user-friendly alternative AdminACL contract that could be designed would have external functions that correspond 1:1 with all core contract admin-callable functions. For example, `updateProjectDescription` could be a function on the AdminACL that forwards the call to the core contract's `updateProjectDescription` function, if `msg.sender` was allowed to call the function. This AdminACL would have the drawback of having a bunch of boilerplate forwarding code for every function, and it would not be able to add support for new functions (e.g. on new minters) without making the ACL upgradeable (or reverting back to the generalized "call with this raw calldata" function, making it not user-friendly). After some discussion with the team, this was not considered a better option than the implementation in the PR.

### CODE REVIEW

**Comment by jakerockland** [comment]
> @jakerockland - would especially appreciate hearing your thoughts on the following. I am totally open to changing to this implementation if we like it more.
> Perhaps the most competitive and still relatively user-friendly alternative AdminACL contract that could be designed would have external functions that correspond 1:1 with all core contract admin-callable functions. For example, `updateProjectDescription` could be a function on the AdminACL that forwards the call to the core contract's `updateProjectDescription` function, if `msg.sender` was allowed to call the function. This AdminACL would have the drawback of having a bunch of boilerplate forwarding code for every function, however it would be backwards-compatible with V0 and V1 contracts.

@ryley-o could we jam on this a bit synchronously and then loop back? I don't think I'm fully groking the implications here.

File: contracts/interfaces/0.8.x/IAdminACLV0.sol
Comment (jakerockland) [inline contracts/interfaces/0.8.x/IAdminACLV0.sol:11]: I think we want this method to also take a smart contract address, for the address of contract that the `_selector` is associated with–unless intention here is to use `msg.sender` in which case we should make this explicitly documented in the interface (though personally I prefer just sending address along explicitly vs. relying on `msg.sender`).

File: contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol
Comment (jakerockland) [inline contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol:26]: We may want to expose a readable field for the ACL checker of type `IAdminACLV0` in the interface itself for introspect-ability reasons–WDYT?

File: contracts/GenArt721CoreV3.sol
Comment (jakerockland) [inline contracts/GenArt721CoreV3.sol:127]: I feel like we may want to write this explicitly onto a variable as well to expose it directly for introspection purposes–though maybe not worth doing, unsure.

File: contracts/GenArt721CoreV3.sol
Comment (jakerockland) [inline contracts/GenArt721CoreV3.sol:210]: Storing an explicit `IAdminACLV0` instance on the contract would also allow us to avoid casting here.

**Review by jakerockland — APPROVED** [review]
This generally LGTM, though added a couple notes on things that I think require a bit more hashing out :)

**Comment by jakerockland** [comment]
cc @laureli

File: contracts/interfaces/0.8.x/IAdminACLV0.sol
Comment (ryley-o) [reply contracts/interfaces/0.8.x/IAdminACLV0.sol:11]: I have a preference of using `msg.sender` here since it is the true source. e.g. you would never be able to accidentally have a typo in a core contract.

That being said, I can see a use case of a contract wanting to check if a sender has approval on some contract other than itself. How do you feel about overloading `allowed` and adding another function to the interface:
```
function allowed(address _sender, bytes4 _selector, address _contract) external returns (bool);
```
... and updating documentation like you suggest?

Comment (jakerockland) [reply contracts/interfaces/0.8.x/IAdminACLV0.sol:11]: I don't think the typo reason is a good reason, as the same could impact the method signature and that part is passed explicitly. 

The overloading approach sounds good to me if we think we do really want to use the `msg.sender` approach, but I am quite reluctant as I just really don't love how non-explicit it is. 

Even with proper documentation it feels a bit obfuscated (taking the philosophy here that methods should ideally be self-documenting, not requiring reading of the implicit assumptions in dev documentation). 

How compelling do we think the benefit of using `msg.sender` is (may be worth some synchronous convo here, I'm a bit torn).

Comment (ryley-o) [reply contracts/interfaces/0.8.x/IAdminACLV0.sol:11]: hmm, yeah chewing on this more, from the perspective of "ideally self documenting", I'm good with adding an additional arg here instead of using `msg.sender`. If I dig deeper into my psyche here, my initial preference for `msg.sender` was probably driven by gas efficiency more than anything else, but realistically, these functions are not in our mint sequence, and readability/conciseness is more important than <1% gas savings.

Thanks for chewing on this one, will update shortly!

Comment (jakerockland) [reply contracts/interfaces/0.8.x/IAdminACLV0.sol:11]: Yeah I think if this intersected with the minting sequence at all I would definitely feel differently here, but given that these are just for administration method calls, and will have a very low % impact on that (we should definitely still be mindful of _high gas impact_ changes to administration methods) would definitely prefer to optimize for readability in favor of a slight gas trade off. SGTM ser 🙏 TY for the back and forth.

File: contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol
Comment (ryley-o) [reply contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol:26]: yes good call. will update!
Side note, I definitely had a blind spot here with respect to "internalType" return types in ABIs post solidity v5.<something> (ref: https://ethereum.stackexchange.com/questions/76953/what-is-the-purpose-of-internaltype-now-generated-by-the-solidity-compiler-in)

File: contracts/GenArt721CoreV3.sol
Comment (ryley-o) [reply contracts/GenArt721CoreV3.sol:127]: agreed - updated to override and wrap `_transferOwnership` to also update public variable `IAdminACLV0 public adminACLContract` whenever ownership is transferred.

File: contracts/GenArt721CoreV3.sol
Comment (ryley-o) [reply contracts/GenArt721CoreV3.sol:210]: yes, updated 🎉

File: contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol
Comment (ryley-o) [reply contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol:28]: (updated)

File: contracts/interfaces/0.8.x/IAdminACLV0.sol
Comment (ryley-o) [reply contracts/interfaces/0.8.x/IAdminACLV0.sol]: (updated)

**Comment by ryley-o** [comment]
@jakerockland - I think I've addressed all conversations above - prefer to have you resolve your comments since some discussion on this one if you have the time 🙏 thanks ser!

File: contracts/GenArt721CoreV3.sol
Comment (jakerockland) [reply contracts/GenArt721CoreV3.sol:130]: Fantastic–TYVM! :)

File: contracts/GenArt721CoreV3.sol
Comment (jakerockland) [inline contracts/GenArt721CoreV3.sol:225]: This check probably warrants a @dev comment above or an inline comment for improved readability.

**Review by jakerockland — APPROVED** [review]
One additional minor comment but updated version LGTM @ryley-o 

Would love one second set of eyes before merge given the sensitivity here, but no strong opinion on who that is (just nice to have another sanity check).

File: contracts/GenArt721CoreV3.sol
Comment (jakerockland) [reply contracts/GenArt721CoreV3.sol]: LGTM! Thank you :)

File: contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol
Comment (jakerockland) [reply contracts/interfaces/0.8.x/IGenArt721CoreContractV3.sol:28]: 🙌 🙌 🙌 

>  Side note, I definitely had a blind spot here with respect to "internalType" return types in ABIs post-solidity v5.something (ref: https://ethereum.stackexchange.com/questions/76953/what-is-the-purpose-of-internaltype-now-generated-by-the-solidity-compiler-in)

Ah interesting–that was a blind spot for me as well. 😅

File: contracts/interfaces/0.8.x/IAdminACLV0.sol
Comment (jakerockland) [reply contracts/interfaces/0.8.x/IAdminACLV0.sol]: LGTM 🙌  🙌  🙌

**Review by yoshiwarab — APPROVED** [review]

File: contracts/GenArt721CoreV3.sol
Comment (ryley-o) [reply contracts/GenArt721CoreV3.sol:228]: Good callout! updated!

**Pull request merged**
